### PR TITLE
XTypes Dynamic: Fix dt_map_ race

### PR DESCRIPTION
**Pull request by ClaytonCalabrese**
There was a race between the populating of the dt_map_ and the checking of the map.

### CODE REVIEW

File: dds/DCPS/RecorderImpl.cpp [hunk 1]
```diff
@@ -262,6 +262,20 @@ void RecorderImpl::notify_subscription_lost(const WriterIdSeq&)
 {
 }
 
+void
+RecorderImpl::add_to_dt_map(PublicationId pub_id, const XTypes::TypeIdentifier& ti)
```
Comment (jrw972) [inline dds/DCPS/RecorderImpl.cpp:266]:
```suggestion
RecorderImpl::add_to_dynamic_type_map(PublicationId pub_id, const XTypes::TypeIdentifier& ti)
```

File: dds/DCPS/RecorderImpl.h [hunk 2]
```diff
@@ -96,6 +96,8 @@ class OpenDDS_Dcps_Export RecorderImpl
 
   void remove_all_associations();
 
+  void add_to_dt_map(PublicationId pub_id, const XTypes::TypeIdentifier& ti);
```
Comment (jrw972) [inline dds/DCPS/RecorderImpl.h:99]:
```suggestion
  void add_to_dynamic_type_map(PublicationId pub_id, const XTypes::TypeIdentifier& ti);
```

File: dds/DCPS/RTPS/Sedp.cpp [hunk 3]
```diff
@@ -7339,6 +7340,17 @@ void Sedp::match_continue(const GUID_t& writer, const GUID_t& reader)
...
     } else if (call_reader) {
       ReaderAssociationRecord rac(drr, reader, wa);
+      if (use_xtypes_complete_ && reader_type_info->complete.typeid_with_size.type_id.kind() == XTypes::TK_NONE) {
+        // Reader is a local recorder using complete types
+        DCPS::DataReaderCallbacks_rch lock = rac.callbacks_.lock();
+        OpenDDS::DCPS::RecorderImpl* ri = dynamic_cast<OpenDDS::DCPS::RecorderImpl*>(lock.in());
+        if (ri) {
+          XTypes::TypeInformation type_info;
+          if (XTypes::deserialize_type_info(type_info, rac.writer_association_.serializedTypeInfo)) {
+            ri->add_to_dt_map(rac.writer_id(), type_info.complete.typeid_with_size.type_id);
```
Comment (jrw972) [inline dds/DCPS/RTPS/Sedp.cpp:7350]: I'm assuming that this not a "complete" solution.

**Review by jrw972 — CHANGES_REQUESTED** [review]

**Review by jrw972 — APPROVED** [review]

File: dds/DCPS/RecorderImpl.cpp [hunk 4]
```diff
@@ -262,6 +262,20 @@ void RecorderImpl::notify_subscription_lost(const WriterIdSeq&)
 {
 }
 
+void
+RecorderImpl::add_to_dynamic_type_map(PublicationId pub_id, const XTypes::TypeIdentifier& ti)
```
Comment (mitza-oci) [inline dds/DCPS/RecorderImpl.cpp:266]:
```suggestion
RecorderImpl::add_to_dynamic_type_map(const PublicationId& pub_id, const XTypes::TypeIdentifier& ti)
```

File: dds/DCPS/RecorderImpl.h [hunk 5]
```diff
@@ -96,6 +96,8 @@ class OpenDDS_Dcps_Export RecorderImpl
 
   void remove_all_associations();
 
+  void add_to_dynamic_type_map(PublicationId pub_id, const XTypes::TypeIdentifier& ti);
```
Comment (mitza-oci) [inline dds/DCPS/RecorderImpl.h:99]:
```suggestion
  void add_to_dynamic_type_map(const PublicationId& pub_id, const XTypes::TypeIdentifier& ti);
```

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Review by mitza-oci — APPROVED** [review]

**Pull request merged**